platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,482 @@
+                startNewCompactionFile(compactionLevelInProgress.getAndSet(0));
+            }
+        } finally {
+            snapshotCompactionLock.release();
+        }
+    }
+
+    /**
+     * Compact (merge) all files that match the given filter.
+     *
+     * @param reportDurationMetricFunction function to report how long compaction took, in ms
+     * @param reportSavedSpaceMetricFunction function to report how much space was compacted, in Mb
+     * @throws IOException if there was a problem merging
+     * @throws InterruptedException if the merge thread was interupted
+     */
+    @SuppressWarnings("unchecked")
+    public void compact(
+            CASableLongIndex index,
+            @Nullable final BiConsumer<CompactionType, Long> reportDurationMetricFunction,
+            @Nullable final BiConsumer<CompactionType, Double> reportSavedSpaceMetricFunction)
+            throws IOException, InterruptedException {
+        Instant timestamp = Instant.now(clock);
+
+        final UnaryOperator<List<DataFileReader<?>>> filesToCompactFilter;
+        final CompactionType targetCompactionLevel;
+        final String storeName = dataFileCollection.getStoreName();
+        final List<? extends DataFileReader<?>> allCompactableFiles = dataFileCollection.getAllCompletedFiles();
+
+        if (isTimeForFullCompaction(timestamp)) {
+            lastFullCompact = timestamp;
+            /* Filter nothing during a full merge */
+            filesToCompactFilter = dataFileReaders -> dataFileReaders;
+            targetCompactionLevel = FULL;
+            logger.info(MERKLE_DB.getMarker(), "[{}] Starting Large Merge", storeName);
+        } else if (isTimeForMediumCompaction(timestamp)
+                ||
+                // This is a temporary solution for the intense load (like hammer tests) where we create too many files.
+                // It will be removed once the solution comes with #7501 is implemented. The plan is to have a
+                // compaction list.
+                // That is, before we run a compaction we create a list of files from (possibly) all levels to be
+                // compacted.
+                // See https://www.notion.so/swirldslabs/Compaction-Improvements-247726614d924fbaa34aa82a157a2f20 for
+                // details
+
+                (readersOfLevel(SMALL)
+                                .apply((List<DataFileReader<?>>) allCompactableFiles)
+                                .size()
+                        > MAX_FIRST_LEVEL_FILES_ALLOWED)) {
+            lastMediumCompact = timestamp;
+            filesToCompactFilter = readersOfLevel(SMALL);
+            targetCompactionLevel = MEDIUM;
+            logger.info(MERKLE_DB.getMarker(), "[{}] Starting Medium Merge", storeName);
+        } else {
+            filesToCompactFilter = readersOfLevel(NO_COMPACTION);
+            targetCompactionLevel = SMALL;
+            logger.info(MERKLE_DB.getMarker(), "[{}] Starting Small Merge", storeName);
```

Comment:
Logger level should be `debug` here (and above). Too many log messages for empty compactions.